### PR TITLE
Implement inventory functionality.

### PR DESCRIPTION
**Pull request by darko-poljak**
Hello!

I implemented simple inventory.

The idea is to have simple tagging inventory.
There is a list of hosts and for each host there are tags.
Inventory database is a set of files under inventory db base dir.
Filename equals hostname.
Each file contains tags for hostname (defined as file's name).
Each tag is on its own line.

With cdist inventory interface you can list hosts/tags,
add-host, add-tag, del-host, del-tag.

config now has new options, -t and -a.
-t means that host specifies tag - all hosts that have specified tag(s) is selected. -a means that selected hosts must contain ALL specified tag(s). There is also equal -a option for inventory list command.

Tags have no values, as tags are just tags.
Tag name - tag value would in this context mean that host has two tags,
and it is selected it both tags are present (-a option).

This inventory is KISS cdist built-in inventory db.

@telmich @asteven  Please take a look and give your opinion.

### CODE REVIEW

**Comment by telmich** [comment]
Looks good to me - maybe move it to "betainventory" and add a policy file + blog post?

**Comment by darko-poljak** [comment]
@telmich Renamed to betainventory, but only in the minimum manner, in command name, NOTE in help and in man pages. Also added bash/zsh completions for betainventory.

**Comment by darko-poljak** [comment]
@asteven We have a request to also add support to use tags in manifest as we currently use host with env var __target_host. @telmich said that you might have something implemented/written. Do you?

**Comment by 4nd3r** [comment]
what's the current situation with this pull request?

**Comment by telmich** [comment]
Hello 4nd3r,

we still have it pending for some further testing - are you open for giving it a try?

Best regards,

Nico

**Comment by 4nd3r** [comment]
yeah, sure, i can test it. 

into what version should I merge `darko-poljak:inventory` changes?

**Comment by telmich** [comment]
@4nd3r you can use the branch directly w/o merging; however merging into master (if you prefer) should also be possible

**Comment by darko-poljak** [comment]
@4nd3r You can clone my cdist fork and checkout inventory branch and then check it. I have just merged it with ungleich/master.

**Comment by 4nd3r** [comment]
if would like to set `~/.cdist/inventory` as INVENTORY_DIR. just like all my types, explorers and manifests are in `~/.cdist/`.

using `cdist inventory list -I ~/.cdist/inventory/` is too much typing and symlinking isn't proper solution imho.

**Comment by darko-poljak** [comment]
@4nd3r @telmich What do you think? If HOME env var is present then inventory_dir, by default, is $HOME/.cdist/inventory, otherwise current default value will be used. I will also add env var override to se inventory dir.
Ok?

**Comment by 4nd3r** [comment]
yes, using similar principles as cdist[1] would be nice.

[1] http://www.nico.schottelius.org/software/cdist/man/latest/cdist-reference.html?highlight=paths

**Comment by darko-poljak** [comment]
@4nd3r I made changes.
Inventory dir is set by the following (when rule is satisfied then it's directory is used):
1. inventory_dir argument
2. CDIST_INVENTORY_DIR env var if set
3. ~/.ctt/inventory if HOME env var is set
4. distribution inventory directory.

So, if no argument is passed and CDIST_INVENTORY_DIR is not set then if HOME env var is set
the ~/.ctt/inventory is used.
Distribution inventory directory is fallback if none of the previous 3 is satisfied.

**Comment by telmich** [comment]
Are you confusing ctt and cdist here? :-)

**Comment by darko-poljak** [comment]
@telmich Ohhh... :) obviously too little sleep. In code and functioning .cdist is used. But in comments and docs I wrote .ctt. I am going to fix this.

**Comment by 4nd3r** [comment]
@darko-poljak thanks! I'll test inventory functionality next week.

**Comment by 4nd3r** [comment]
```
/opt/cdist# git describe
4.3.2-174-gce2b6ab

# cdist inventory list -t testtag -b | wc -l
9

# cdist config -d -t testtag -i .cdist/manifest/test -b
DEBUG: cdist: Namespace(beta=True, command='config', conf_dir=None, debug=True, dry_run=False, func=<bound method type.commandline of <class 'cdist.config.Config'>>, has_all_tags=False, host=['testtag'], hostfile=None, inventory_dir=None, jobs=None, manifest='.cdist/manifest/test', out_path=None, parallel=False, remote_copy=None, remote_exec=None, tag=True, verbose=False)
INFO: cdist: version 4.3.2-174-gce2b6ab
DEBUG: inventory: Init db: /opt/cdist/cdist/inventory
INFO: inventory: Listing by tag(s)
DEBUG: inventory: param_tags: {'testtag'}
INFO: cdist: Total processing time for 0 host(s): 0.006699085235595703
````

but adding `-I .cdist/inventory` works.

**Comment by darko-poljak** [comment]
Oh, I have to review config command for setting default inventory dir.

Darko

**All work and no play makes Jack a dull boy**

**Comment by darko-poljak** [comment]
@4nd3r I have fixed default inventory dir in config command.

**Comment by 4nd3r** [comment]
behaviour of `-a` stays bit hazy. how should it work?

i would like a way to execute in all hosts, at least that's my expectation with `-a`.

**Comment by darko-poljak** [comment]
@4nd3r -a for config command means that all hosts will be listed that contain specified tag (-t).
Without -a and with -t specified config will configure all hosts that contain any of specified tags.
config -t tag1 tag2   ---> configure hosts that have any of the specified tags, for example, if host A has tags: tag1, tag3, tag4 it will be configured
config -t -a tag1 tag2  --> configure hosts that have all specified tags, for example, if host A has tags: tag1, tag3, tag4 it will not be configured.

I can add new option, for example, -A: if -A is specified along -t then all hosts will be configure, regardless specified tags at command line.
What do you think?

**Comment by 4nd3r** [comment]
ahaa, okay, now I get it. tho, parameters order is bit strange.
i was trying `cdist config -a -t tag1 tag2` and `cdist config -t tag1 tag2 -a`.
`-A` or just `-a` without `-t` for selecting all hosts in inventory would be cool, thanks!

**Comment by darko-poljak** [comment]
@4nd3r I added -A/--all-tagged option.
`cdist config -b -A`
should now configure all hosts present in inventory db.
Give it a try, I haven't had much time to test it in detail.
Thanks.

File: docs/src/man1/cdist.rst [hunk 1]
```diff
@@ -507,6 +514,9 @@ EXAMPLES
     # Configure hosts from inventory with all specified tags
     % cdist config -b -t -a web dynamic
 
+    # Configure all hosts from inventory db
+    $ cdist -b -A
```
Comment (4nd3r) [inline docs/src/man1/cdist.rst]: did you mean `cdist config -b -A`?

Comment (darko-poljak) [inline docs/src/man1/cdist.rst]: @4nd3r Yes, thanks for noticing.

**Comment by darko-poljak** [comment]
@4nd3r I have fixed examples in docs.

**Comment by 4nd3r** [comment]
@darko-poljak thanks for your hard work! i'll continue testing.

**Comment by darko-poljak** [comment]
Hi, @4nd3r ,
how is testing going?

**Comment by 4nd3r** [comment]
got bitten by `__package_upgrade_all` when doing blind upgrades - see #494.
but that's not related to inventory, which, after the fixes, works great!
managing many hosts became much more easier.

thanks!

**Comment by darko-poljak** [comment]
@4nd3r Can you continue testing inventory from ungleich/beta branch? There are also other new features like trigger and preos. I think I will close this PR and we continue in beta branch.
@telmich Do you agree?

**Comment by telmich** [comment]
@darko-poljak agreed

**Pull request closed without merging**